[Merged by Bors] - feat(data/set/intervals/instances): algebraic instances for unit intervals

stuart-presnell opened the pull request:
Proving `cancel_comm_monoid_with_zero (Icc 0 1)`, `comm_semigroup (Ico 0 1)`, `comm_monoid (Ioc 0 1)`, and `comm_semigroup (Ioo 0 1)` for suitably structured underlying type `α`.

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

---

This replaces #14288.  I decided I'd tried to do too much in that PR by also incorporating material from `topology/unit_interval`.  So I've started over with just the basics, and left that extra job as a TODO.

cc @eric-wieser @YaelDillies @urkud  who reviewed #14288

This PR originally stems from a comment by @YaelDillies in [this](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/nat.20powers.20of.20probabilities/near/282480844) Zulip thread.


<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
Create instances.lean

stuart-presnell committed:
Add zero/one lemmas for Ico with [nontrivial α]

stuart-presnell committed:
Add zero/one lemmas for Ioc with [nontrivial α]

stuart-presnell committed:
Fix copy & paste error

stuart-presnell committed:
Fix overlong line

stuart-presnell committed:
import `.intervals.instances` in `unit_interval`

stuart-presnell committed:
Fix overlong lines

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+import algebra.order.ring
+import algebra.group_power.order
+
+import data.set.intervals.proj_Icc
```
```suggestion
import algebra.group_power.order
import data.set.intervals.proj_Icc
```

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+/-- like `nonneg`, but with the inequality in `Icc (0:α) 1`. -/
+lemma nonneg' {t : Icc (0:α) 1} : 0 ≤ t := t.2.1
+/-- like `le_one`, but with the inequality in `Icc (0:α) 1`. -/
+lemma le_one' {t : Icc (0:α) 1} : t ≤ 1 := t.2.2
```
How about:
```suggestion
lemma coe_nonneg (x : Icc (0:α) 1) : 0 ≤ (x : α) := x.2.1
lemma coe_le_one (x : Icc (0:α) 1) : (x : α) ≤ 1 := x.2.2

/-- like `coe_nonneg`, but with the inequality in `Icc (0:α) 1`. -/
lemma nonneg {t : Icc (0:α) 1} : 0 ≤ t := t.2.1
/-- like `coe_le_one`, but with the inequality in `Icc (0:α) 1`. -/
lemma le_one {t : Icc (0:α) 1} : t ≤ 1 := t.2.2
```

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+⟨one_sub_mem, λ h, (sub_sub_cancel 1 t) ▸ one_sub_mem h⟩
+
+lemma one_minus_nonneg (x : Icc (0:β) 1) : 0 ≤ 1 - (x : β) := by simpa using x.2.2
+lemma one_minus_le_one (x : Icc (0:β) 1) : 1 - (x : β) ≤ 1 := by simpa using x.2.1
```
```suggestion
lemma one_sub_nonneg (x : Icc (0:β) 1) : 0 ≤ 1 - (x : β) := by simpa using x.2.2
lemma one_sub_le_one (x : Icc (0:β) 1) : 1 - (x : β) ≤ 1 := by simpa using x.2.1
```

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+lemma le_one (x : Ioc (0:α) 1) : (x : α) ≤ 1 := x.2.2
+
+/-- like `le_one`, but with the inequality in `Ioc (0:α) 1`. -/
+lemma le_one' [nontrivial α] {t : Ioc (0:α) 1} : t ≤ 1 := t.2.2
```
As above:
```suggestion
lemma coe_pos (x : Ioc (0:α) 1) : 0 < (x : α) := x.2.1
lemma coe_le_one (x : Ioc (0:α) 1) : (x : α) ≤ 1 := x.2.2

/-- like `coe_le_one`, but with the inequality in `Ioc (0:α) 1`. -/
lemma le_one [nontrivial α] {t : Ioc (0:α) 1} : t ≤ 1 := t.2.2
```

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+lemma lt_one (x : Ico (0:α) 1) : (x : α) < 1 := x.2.2
+
+/-- like `nonneg`, but with the inequality in `Ico (0:α) 1`. -/
+lemma nonneg' [nontrivial α] {t : Ico (0:α) 1} : 0 ≤ t := t.2.1
```
As above:
```suggestion
lemma coe_nonneg (x : Ico (0:α) 1) : 0 ≤ (x : α) := x.2.1
lemma coe_lt_one (x : Ico (0:α) 1) : (x : α) < 1 := x.2.2

/-- like `coe_nonneg`, but with the inequality in `Ico (0:α) 1`. -/
lemma nonneg [nontrivial α] {t : Ico (0:α) 1} : 0 ≤ t := t.2.1
```

Vierkantor commented on src/topology/unit_interval.lean:
```diff
@@ -6,6 +6,7 @@ Authors: Patrick Massot, Scott Morrison
 import topology.instances.real
 import topology.algebra.field
 import data.set.intervals.proj_Icc
+import data.set.intervals.instances
```
Presumably we don't need this import yet?
```suggestion
```

Vierkantor commented on src/data/set/intervals/instances.lean:
```diff
@@ -0,0 +1,254 @@
+lemma coe_ne_zero {x : Icc (0:α) 1} : (x : α) ≠ 0 ↔ x ≠ 0 :=
+not_iff_not.mpr coe_eq_zero
+
+@[simp] lemma mk_one (h : (1 : α) ∈ Icc (0 : α) 1) : (⟨1, h⟩ : Icc (0:α) 1) = 1 := rfl
```
Could you move this next to `mk_zero`?

Vierkantor reviewed (approved):
Thanks!

bors d+

stuart-presnell committed:
Edit imports

stuart-presnell committed:
Move `mk_one`

stuart-presnell committed:
Add `coe`s, delete primes

stuart-presnell committed:
Change `minus` to `sub`

stuart-presnell committed:
Merge branch 'master' into SP_interval_instances
Fix `positivity` tactic

stuart-presnell committed:
Edit lemmas to use `set.Icc` counterparts

Just to see what the `simp_nf` linter says about this.  Presumably the eventual solution will be to delete these lemmas and directly use the `set.Icc` versions wherever these were used?

YaelDillies commented on src/topology/unit_interval.lean:
```diff
@@ -76,7 +75,7 @@ instance : nonempty I := ⟨0⟩
 instance : has_mul I := ⟨λ x y, ⟨x * y, mul_mem x.2 y.2⟩⟩
 
-@[simp, norm_cast] lemma coe_mul {x y : I} : ((x * y : I) : ℝ) = x * y := rfl
+@[simp, norm_cast] lemma coe_mul {x y : I} : ((x * y : I) : ℝ) = x * y := Icc.coe_mul x y
```
You should keep those `rfl` so that `dsimp` learns about them.

stuart-presnell replied:
These lemmas have all been generalised in `data/set/intervals/instances` to `set.Icc`, and those whose proofs were `rfl` here are also `rfl` in the generalisation so I assume this shouldn't change anything for `dsimp`.  So I'm deleting them from `topology/unit_interval` and now I'm waiting for CI to confirm that this doesn't break anything.

YaelDillies replied:
Yeah, that sounds better.

stuart-presnell committed:
Delete simp lemmas that have been generalised

These lemmas have all been generalised in `data/set/intervals/instances` to `set.Icc`, and those whose proofs were `rfl` here are also `rfl` in the generalisation so I assume this shouldn't change anything for `dsimp`.
